File: forms/lifeguard-certification-checklist/full-stack-with-loco-tera-htmx-alpine/src/engine/lifeguard_grader.rs
// Lifeguard Competency Verification grader. Pure functions: take an
// `AssessmentData` object, evaluate each declarative rule in the registry,
// and return the outcome plus the audit trail of fired rules.
//
// RLSS UK NPLQ outcome logic:
//   - Any critical-competency rule firing "no" -> overall Fail.
//   - With no critical failures, 0 deficiencies -> Pass.
//   - With no critical failures and 1-2 non-critical deficiencies ->
//     Needs Development.
//   - With no critical failures and >2 non-critical deficiencies ->
//     Needs Development (high-deficiency case surfaced as a flag for the
//     examiner to review).
//   - Items the examiner has not assessed ("na" or "") do not count as
//     deficiencies.

use super::flagged_issues::detect_additional_flags;
use super::rules::lifeguard_rules;
use super::types::{AssessmentData, FiredRule, GradingResult, Outcome};

pub const DEFICIENCY_LIMIT_FOR_PASS: usize = 0;
pub const DEFICIENCY_LIMIT_FOR_NEEDS_DEVELOPMENT: usize = 2;

/// Evaluate every lifeguard rule in the registry, classify the outcome, and
/// return the full grading result.
pub fn grade(data: &AssessmentData) -> GradingResult {
    let timestamp = chrono::Utc::now().to_rfc3339();
    let rules = lifeguard_rules();

    let mut fired_rules: Vec<FiredRule> = Vec::new();
    let mut critical_failures: Vec<FiredRule> = Vec::new();
    let mut deficiencies: Vec<FiredRule> = Vec::new();
    let mut answered_count: u32 = 0;

    for rule in &rules {
        let status = (rule.evaluate)(data);

        let entry = FiredRule {
            id: rule.id.to_string(),
            step: rule.step,
            category: rule.category.to_string(),
            description: rule.label.to_string(),
            critical: rule.critical,
            status: status.clone(),
        };
        fired_rules.push(entry.clone());

        if status == "yes" || status == "no" {
            answered_count += 1;
        }

        if status == "no" {
            if rule.critical {
                critical_failures.push(entry);
            } else {
                deficiencies.push(entry);
            }
        }
    }

    let outcome: Outcome = if !critical_failures.is_empty() {
        "fail".to_string()
    } else if answered_count == 0 {
        // Nothing assessed yet — surface a neutral "fail" on submission so the
        // examiner is prompted to actually mark items.
        "fail".to_string()
    } else if deficiencies.len() == DEFICIENCY_LIMIT_FOR_PASS {
        "pass".to_string()
    } else if deficiencies.len() <= DEFICIENCY_LIMIT_FOR_NEEDS_DEVELOPMENT {
        "needs-development".to_string()
    } else {
        "needs-development".to_string()
    };

    let additional_flags = detect_additional_flags(data, &critical_failures, &deficiencies);

    GradingResult {
        outcome,
        critical_failures,
        deficiencies,
        fired_rules,
        additional_flags,
        answered_count,
        total_rules: rules.len() as u32,
        timestamp,
    }
}

File: forms/lifeguard-certification-checklist/front-end-with-svelte/src/lib/engine/lifeguard-grader.ts
// Lifeguard Competency Verification grader. Pure functions: take an
// `AssessmentData` object, evaluate each declarative rule in the registry,
// and return the outcome plus the audit trail of fired rules.
//
// RLSS UK NPLQ outcome logic:
//   - Any critical-competency rule firing 'no' -> overall Fail.
//   - With no critical failures and 0 deficiencies -> Pass.
//   - With no critical failures and >=1 non-critical deficiency ->
//     Needs Development (the candidate may remediate and re-present rather
//     than outright Fail when no critical breach occurs).
//   - Items the examiner has not assessed ('na' or '') do not count as
//     deficiencies.

import type { AssessmentData, FiredRule, GradingResult, Outcome } from './types';
import { lifeguardRules } from './rules';
import { detectAdditionalFlags } from './flagged-issues';

export const DEFICIENCY_LIMIT_FOR_PASS = 0;
export const DEFICIENCY_LIMIT_FOR_NEEDS_DEVELOPMENT = 2;

/**
 * Evaluate every lifeguard rule in the registry, classify the outcome, and
 * return the full audit trail plus the examiner flags.
 */
export function gradeLifeguard(data: AssessmentData): GradingResult {
	const firedRules: FiredRule[] = [];
	const criticalFailures: FiredRule[] = [];
	const deficiencies: FiredRule[] = [];
	let answeredCount = 0;

	for (const rule of lifeguardRules) {
		let status: FiredRule['status'] = '';
		try {
			status = rule.evaluate(data);
		} catch (e) {
			console.warn(`Lifeguard rule ${rule.id} evaluation failed:`, e);
			status = '';
		}

		const entry: FiredRule = {
			id: rule.id,
			step: rule.step,
			category: rule.category,
			description: rule.label,
			critical: rule.critical,
			status
		};
		firedRules.push(entry);

		if (status === 'yes' || status === 'no') answeredCount++;

		if (status === 'no') {
			if (rule.critical) {
				criticalFailures.push(entry);
			} else {
				deficiencies.push(entry);
			}
		}
	}

	let outcome: Outcome = '';
	if (criticalFailures.length > 0) {
		outcome = 'fail';
	} else if (answeredCount === 0) {
		// Nothing assessed yet — surface a neutral "fail" on submission so the
		// examiner is prompted to actually mark items, but don't claim success.
		outcome = 'fail';
	} else if (deficiencies.length === DEFICIENCY_LIMIT_FOR_PASS) {
		outcome = 'pass';
	} else {
		// One or more non-critical deficiencies but no critical breach: the
		// candidate needs targeted development before recertification.
		outcome = 'needs-development';
	}

	const additionalFlags = detectAdditionalFlags(data, { criticalFailures, deficiencies });

	return {
		outcome,
		criticalFailures,
		deficiencies,
		firedRules,
		additionalFlags,
		answeredCount,
		totalRules: lifeguardRules.length,
		timestamp: new Date().toISOString()
	};
}
